package ZPMeta;

enum StatusCode {
  OK = 0;
  NOTFOUND = 1;
  ERROR = 2;
}

enum Type {
  PING = 1;
  PULL = 2;
  INIT = 3;
  SETMASTER = 4;
  ADDSLAVE = 5;
  REMOVESLAVE = 6;
  LISTTABLE = 7;
  LISTNODE = 8;
  LISTMETA = 9;
}

enum PState {
  ACTIVE = 1;
  STUCK = 2;
}

message Node {
  required string ip = 1;
  required int32 port = 2;
}

message NodeStatus {
  required Node node = 1;
  required int32 status = 2;
}

message Nodes {
  repeated NodeStatus nodes = 1;
}

message MetaNodes {
  repeated Node followers = 1;
  optional Node leader = 2;
}

message Partitions {
  required int32 id = 1;
  required PState state = 2;
  required Node master = 3;
  repeated Node slaves = 4;
}

message TableName {
  repeated string name = 1;
}

message Table {
  required string name = 1;
  repeated Partitions partitions = 2;
}

message BasicCmdUnit {
  required string name = 1;
  required int32 partition = 2;
  required Node node = 3;
}

message SyncOffset {
  required string name = 1;
  required int32 partition = 2;
  required int32 filenum = 3;
  required int64 offset = 4;
}

message MetaCmd {

  required Type type = 1;

  // Ping 
  message Ping {
    required int32 version = 1;
    required Node node = 2;
    repeated SyncOffset offset = 3;
  }
  optional Ping ping = 2;

  // Pull
  message Pull {
    optional Node node = 1;
    optional string name = 2;
  }
  optional Pull pull = 3;

  // Init
  message Init {
    required string name = 1;
    required int32 num = 2;
  }
  optional Init init = 4;

  message SetMaster{
    required BasicCmdUnit basic = 1;
  }
  optional SetMaster set_master = 5;

  message AddSlave {
    required BasicCmdUnit basic = 1;
  }
  optional AddSlave add_slave = 6;

  message RemoveSlave {
    required BasicCmdUnit basic = 1;
  }
  optional RemoveSlave remove_slave = 7;
}

message MetaCmdResponse {

  required Type type = 1;
  required StatusCode code = 2;
  optional string msg = 3;

  // Ping
  message Ping {
    required int32 version = 1;
    optional bool need_offset = 2;
    optional int32 interval = 3; 
  }
  optional Ping ping = 4;

  // Pull
  message Pull {
    required int32 version = 1;
    repeated Table info = 2;
  }
  optional Pull pull = 5;

  // ListTable 
  message ListTable {
    optional TableName tables = 1;
  }
  optional ListTable list_table = 6;

  // ListNode
  message ListNode {
    optional Nodes nodes = 1;
  }
  optional ListNode list_node = 7;

  // ListMeta
  message ListMeta {
    required MetaNodes nodes = 1;  
  }
  optional ListMeta list_meta = 8;
}
